File: tests/fixtures/corpus-full/python/services.py
from typing import List, Optional
from models import User, Product, Order
from db import Database
from auth import AuthManager


class UserService:
    def __init__(self, db: Database, auth: AuthManager) -> None:
        self.db = db
        self.auth = auth

    def register(self, username: str, email: str, password: str) -> User:
        user = User(id=hash(username), username=username, email=email)
        self.db.save_user(user)
        return user

    def get_profile(self, token: str) -> Optional[User]:
        return self.auth.get_user(token)

    def deactivate_account(self, token: str) -> bool:
        user = self.auth.get_user(token)
        if user is None:
            return False
        user.deactivate()
        self.db.save_user(user)
        return True


class ProductService:
    def __init__(self, db: Database) -> None:
        self.db = db

    def list_available(self) -> List[Product]:
        return []

    def get_by_id(self, product_id: int) -> Optional[Product]:
        return self.db.find_product(product_id)

    def restock(self, product_id: int, qty: int) -> bool:
        product = self.db.find_product(product_id)
        if product is None:
            return False
        product.stock += qty
        self.db.save_product(product)
        return True


class OrderService:
    def __init__(self, db: Database, product_svc: ProductService) -> None:
        self.db = db
        self.product_svc = product_svc

    def create_order(self, user: User) -> Order:
        order = Order(id=hash(user.username), user=user)
        self.db.save_order(order)
        return order

    def add_item(self, order_id: int, product_id: int, qty: int = 1) -> bool:
        order = self.db.find_order(order_id)
        product = self.product_svc.get_by_id(product_id)
        if order is None or product is None:
            return False
        order.add_item(product, qty)
        self.db.save_order(order)
        return True

File: tests/fixtures/corpus-full/rust/services.rs
use crate::db::Database;
use crate::auth::AuthManager;
use crate::models::{User, Product, Order, OrderStatus};

pub struct UserService {
    db: Database,
    auth: AuthManager,
}

impl UserService {
    pub fn new(db: Database, auth: AuthManager) -> Self {
        Self { db, auth }
    }

    pub fn register(&self, username: &str, email: &str) -> User {
        let user = User::new(0, username, email);
        self.db.save_user(&user);
        user
    }

    pub fn get_profile(&self, token: &str) -> Option<User> {
        self.auth.get_user(token)
    }

    pub fn deactivate(&self, token: &str) -> bool {
        let mut user = self.auth.get_user(token)?;
        user.deactivate();
        self.db.save_user(&user);
        Some(()).is_some()
    }
}

pub struct ProductService {
    db: Database,
}

impl ProductService {
    pub fn new(db: Database) -> Self {
        Self { db }
    }

    pub fn get_by_id(&self, id: u64) -> Option<Product> {
        self.db.find_product(id)
    }

    pub fn restock(&self, id: u64, qty: u32) -> bool {
        let mut product = self.db.find_product(id)?;
        product.stock += qty;
        self.db.save_product(&product);
        Some(()).is_some()
    }
}

pub struct OrderService {
    db: Database,
    product_service: ProductService,
}

impl OrderService {
    pub fn new(db: Database, product_service: ProductService) -> Self {
        Self { db, product_service }
    }

    pub fn create(&self, user_id: u64) -> Order {
        let order = Order { id: 0, user_id, items: vec![], total: 0.0, status: OrderStatus::Pending };
        self.db.save_order(&order);
        order
    }
}
